### ISSUE TITLE
router.navigate option to disable URI encoding

### ISSUE THREAD

**Issue by verylongword** [angular/angular#28917]
# 🚀 feature request

### Relevant Package
 This feature request is for @angular/router

### Description
The problem is that router.navigate automatically encodes a URL  ( using encodeURIComponent( .. )  I think),  but this encoding is not idempotent.
For example repeated encodes of a path with parameter segments will duplicate encoding of a '% '
Consequently, navigation from a path constructed from prior navigation will cause disaster when 
decodeURIComponent( .. ) is used to recover the parameter object .

### Describe the solution you'd like
An optional parameter to disable URL encoding.
Then the user can avoid this problem by encoding each parameter just once. 

This  parameter could be an option  in router.navigate,    (easy for a user to notice)

Alternatively a module option in
    RouterModule.forRoot(routes,
    {    enableTracing: true,
      initialNavigation: true,
      relativeLinkResolution: 'corrected',   // Fix Router BUG
      disableUrlEncoding: true                  //  Suggest something like this: but would need documentation
     }

### Describe alternatives you've considered
1.   navigateByUrl:   but I would prefer not to construct absolute URL's ( with possibly unexpected limitations) 
2.   Custom UrlSerializer:  but this seems unreasonably difficult for  an average user to investigate and implement.

**Comment by hezzyg** [comment]
This feature is very important. Current situation is that when url contains query params, the next time you navigate to this url it will get encoded - causing ActivatedRoute to ignore these query params and destroying url-params exist in the url because the encoded q-params are concatenated to the url-param... 

e.g
domain.com/post/234234ewrewrewr**?user=true**

turns into this once you navigate to the same url ->

domain.com/post/234234ewrewrewr**%3Fuser%3Dtrue**?user=true

**Comment by Karamuto** [comment]
Kind of hacky workaround for this:

When providing a custom 'UrlSerializer', you can manipulate the Url tree in the 'serialize' method which is used for the 'match' method later on. 

I am not happy with this solution but it works at least for the time beeing.

**Comment by deshpandeashu** [comment]
Any plan to address this issue?

**Comment by penfold** [comment]
This is giving issues with an email address as a query param.

**Comment by jonathanlie** [comment]
any updates on this?

**Comment by jonathanlie** [comment]
update: we worked around this by using `navigateByUrl`, and it works fine.

**Comment by beezital** [comment]
I've stumbled on this issue but realised that trying to build the URLs myself was inappropriate.
My code now uses:

**for html:** 
`[routerLink]="'/my/path/' + myVar" [queryParams]="{param1: value1, param2: value2}"`

**for component:**
```
var navigateTo = '/my/path/' + myVar;
var navigationExtras = {
	queryParams: { param1: value1, param2: value2 },
	replaceUrl: true, // optional
};
this.router.navigate([navigateTo], navigationExtras);
```

**Comment by kernyden** [comment]
i cannot navigato to something like: 
```HTML
kanbanboard/295?teams=all&queryfilters=[{id: 'status_id', values: [1, 2, 3], operator: 'o'}]
```
it's been transformed to:
```HTML
/kanbanboard/295?teams=all&queryfilters=%5B%7Bid:%20%27status_id%27,%20values:%20%5B1,%202,%203%5D,%20operator:%20%27o%27%7D%5D
```
is there some workaround?

**Comment by beezital** [comment]
@kernyden I am wondering how and where are you decoding the url.
my tests show that your query string (`queryfilters=%5B%7Bid:%20%27status_id%27,%20values:%20%5B1,%202,%203%5D,%20operator:%20%27o%27%7D%5D`) is properly encoded. When reading it with e.g. [route.queryParams](https://angular.io/api/router/ActivatedRoute) it gets automatically decoded:
```
this.route.queryParams.subscribe((queryParams: Params) => {
  if (typeof queryParams.queryfilters != 'undefined') {
    console.log('queryfilters=', queryParams.expandmenu);
  }
});
// ==> queryfilters= [{id: 'status_id', values: [1, 2, 3], operator: 'o'}]
```
https://www.urldecoder.org/
https://en.wikipedia.org/wiki/Query_string

**Comment by kernyden** [comment]
@beezital true, my question might be hardly relevant to the current, but i want to ignore special characters used by browser and redirect to the encoded url in browser's own.

for instance: 
![image](https://user-images.githubusercontent.com/25119196/87717609-59f00200-c7b9-11ea-86b1-e58e07819dc5.png)

**Comment by beezital** [comment]
@kernyden I would say that the URL you show should never be generated nor appear as such.
see **RFC 3986**:
https://tools.ietf.org/html/rfc3986#page-14 _When to Encode or Decode_
https://tools.ietf.org/html/rfc3986#appendix-A
```
reserved      = gen-delims / sub-delims
gen-delims    = ":" / "/" / "?" / "#" / "[" / "]" / "@"
sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
                 / "*" / "+" / "," / ";" / "="
```

**Comment by vtsykun** [comment]
Hi, the same problem. The char `#` converted to %23, but given breaks angular routing  

```
this.router.navigate(['/cases/view/53#comment-1']);
.... 

const appRoutes: Routes = [
...
  { path: 'cases/view/:id', component: CaseViewComponent, data: { title: 'Title' }, canActivate: [AuthGuard]},
...

const id = this.currentRouter.snapshot.paramMap.get('id'); // id != 53  

```

**Comment by mwaibel-go** [comment]
This problem also occurs when you load a particular route in your application when first loading the app. For instance, if you find a link in an email like https://my-ng-app/#/entities/12?subentity=13. When angular tries to load this, the query parameter gets encoded and can’t be accessed via `ActivatedRoute::queryParams`.

I therefore think that a possible solution should not be implemented in a way that only solves in-app navigation (i.e. the addition to `Router::navigate`), but also the suggested route-specific configuration or a Router-wide configuration.

**Comment by fletz** [comment]
Is there any progress on this? I have a path param, which contains a + sign. Which shouldn't even be a problem in path, but it get's encoded to %2B. And if I press F5, the % sign get's encoded again and so forth. Same happens in the app when routing,

**Comment by BDengCapsifi** [comment]
seems still an issue.
i also encountered this the `?` been encoded after 1st refresh then keep encoding `%` when refresh the same page

**Comment by ggarg2** [comment]
+1 I am also facing the same issue.

**Comment by atscott** [comment]
I'm trying to understand the actual issues being encountered here but there aren't really any clear steps to reproduce. Would it be enough to allow configuring which characters are allowed in the path and query when using the default serializer, similar to the feature in [Tanstack Router](https://tanstack.com/router/latest/docs/framework/react/guide/path-params#allowed-characters). We would still explicitly always encode `;`, `(`, `)`, and `=` in the path segments because of named outlets and matrix parameters.
